### ISSUE TITLE
io.wavfile.read unnecessarily raises an error for a bad wav header

### ISSUE THREAD

**Issue by stephenpcook** [scipy/scipy#10835]
When the header of a .wav file gives the wrong filesize `io.wavfile.read` raises `ValueError("Unexpected end of file.")` even if the data has already been successfully read.  An example of this is actually built in as a test in [test_wavfile.py](https://github.com/scipy/scipy/blob/198ebbab1aa5b2e8757b4e07fcea893df594e88c/scipy/io/tests/test_wavfile.py#L87-L91):
```python
def test_read_early_eof():
    for mmap in [False, True]:
        fp = open(datafile('test-44100Hz-le-1ch-4bytes-early-eof.wav'), 'rb')
        assert_raises(ValueError, wavfile.read, fp, mmap=mmap)
        fp.close()
```
but the file [`test-44100Hz-le-1ch-4bytes-early-eof.wav`](https://github.com/scipy/scipy/blob/198ebbab1aa5b2e8757b4e07fcea893df594e88c/scipy/io/tests/data/test-44100Hz-le-1ch-4bytes-early-eof.wav) is successfully read by other programs (such as VLC, Audacity and MATLAB).  This behaviour was introduced in 171e98ad56025fabd1388496eefeb437d9d810b9 as a solution to #6700.

I propose this can be solved by  checking if the data chunk has been read and if so, silently ignoring the filesize from the malformed head (or perhaps with a warning).  I've implemented this in stephenpcook/scipy@4b3b34d26853f0b209ec8734ce083e6253bb557a but since this"breaks" an existing test I wanted to raise this as an issue and get some advice before creating a pull request.

**Comment by stephenpcook** [comment]
This was raised in [this comment](https://github.com/scipy/scipy/issues/6700#issuecomment-367827147) but I don't believe it was followed up.

**Comment by stephenpcook** [comment]
There is still an outstanding issue where a malformed file cannot be read with `mmap=True` - this is mentioned in [this comment](https://github.com/scipy/scipy/issues/6700#issuecomment-255592052) of #6700.

**Comment by larsoner** [comment]
Fixed by #10870
